azure/src/createAzureClient.ts

```diff
@@ -283,3 +297,75 @@ class AllowInsecureConnectionPolicy implements PipelinePolicy {
         return await next(request);
     }
 }
+
+/**
+ * Resolve a default scope from the Resource Manager endpoint or any provided endpoint.
+ * Example: https://management.azure.com/ -> https://management.azure.com/.default
+ */
+function getDefaultScopeFromEndpoint(endpoint?: string): string {
+    let base = endpoint ?? 'https://management.azure.com/';
+    base = base.replace(/\/+$/, '');
+    return `${base}/.default`;
+}
+
+/**
+ * Try to extract an OAuth scope or resource from the raw WWW-Authenticate header.
+ * Falls back to deriving a scope from the endpoint if not present.
+ */
+function resolveScopesFromChallengeOrEndpoint(rawHeader: string, endpoint?: string): string[] {
+    const scopeMatch = /(?:^|[,\s])scope="([^"]+)"/i.exec(rawHeader);
+    if (scopeMatch && scopeMatch[1]) {
+        return [scopeMatch[1]];
+    }
+
+    const resourceMatch = /(?:^|[,\s])resource(?:_id)?=\"?([^\",\s]+)\"?/i.exec(rawHeader);
+    if (resourceMatch && resourceMatch[1]) {
+        let resource = resourceMatch[1].replace(/\/+$/, '');
+        if (!/\.default$/i.test(resource)) {
+            resource = `${resource}/.default`;
+        }
+        return [resource];
+    }
+
+    return [getDefaultScopeFromEndpoint(endpoint)];
+}
+
+/**
+ * A custom bearer policy that pre-authorizes and then retries once on a 401/403 with a WWW-Authenticate challenge.
+ * It passes the raw WWW-Authenticate header to the credential via options.challenge and enables CAE.
+ */
+class AzExtBearerChallengePolicy implements PipelinePolicy {
+    public readonly name = 'AzExtBearerChallengePolicy';
+    private readonly endpoint?: string;
+    private readonly getTokenForChallenge: (scopes: vscode.AuthenticationSessionRequest) => Promise<string | undefined>;
+    private readonly challengeRetryHeader = 'x-azext-challenge-retry';
+
+    public constructor(getTokenForChallenge: (scopes: vscode.AuthenticationSessionRequest) => Promise<string | undefined>, endpoint?: string) {
```

Comment:
Not imperative, but you can save a few lines of code by doing this with the constructor, instead of copying in the values to the private fields:

```suggestion
    public constructor(private readonly getTokenForChallenge: (scopes: vscode.AuthenticationSessionRequest) => Promise<string | undefined>, private readonly endpoint?: string) {
```

Along with eliminating the fields above